### PR TITLE
feat(config): option to configure a helper to lookup the token

### PR DESCRIPTION
**Pull request by klorenz**
I have stored my personal access token with all other secrets in UNIX password store.  
For not having the token stored on the disk, I have implemented the option to let tokens be
looked up from a helper program.

Example:
```ini
[my_config]
url = https://gitlab.server
private_token = helper: ~/bin/pass-helper.sh
api_version = 4
```
whith a helper script ``~/bin/pass-helper.sh``:
```bash
#!/bin/bash
pass show path/to/token | head -n 1
```

### CODE REVIEW

**Comment by klorenz** [comment]
Hi @nejch ,

to fix commitlinting issues, I have squashed my commits and created better commit messages.

File: gitlab/tests/test_config.py
Comment (nejch) [inline gitlab/tests/test_config.py:209]: Thanks for adding tests for this :)

This test is really growing now, it might make sense to split it into a new `test_data_from_helper`. Then you could check: what error the user will see if the subprocess exits with non-zero, if it returns nothing, in case there's nothing given in the config after the prefix, etc.

Most of the other stuff in this class is wrapped in blanket try/except blocks so I think if we don't do it for helpers, it should be tested well.

File: gitlab/config.py
Comment (nejch) [inline gitlab/config.py:160]: I think we should at least warn the user in the docs about `shell=True` or link to https://docs.python.org/3/library/subprocess.html#security-considerations. I guess you added it here so people can use pipes etc.

Also I guess this opens the user to arbitrary commands executed on the machine if someone manages to put a config file with precedence or in `PYTHON_GITLAB_CFG` env var. @max-wittig WDYT about using `shell=True` and this 2nd point here, does it matter? :thinking:

**Review by nejch — CHANGES_REQUESTED** [review]
Thanks a lot for this feature, this is really really neat :) People have requested native keyring support but I think this might be even more flexible as it can be used with any backend/CLI client like vault etc.

I've just added a few comments, sorry it's just a very user-facing feature so I thought it'd be good to agree on the approach from the start.

Also I think this deserves a mention in the docs, would you mind adding a section in https://github.com/python-gitlab/python-gitlab/blob/master/docs/cli-usage.rst?

File: gitlab/config.py
Comment (klorenz) [reply gitlab/config.py:160]: We could also rather run a simple command and leave it up to the user to create a shell script.  If I remember right, this is the behavior from docker.

File: gitlab/config.py
Comment (klorenz) [reply gitlab/config.py]: Done this.

File: gitlab/config.py
Comment (klorenz) [reply gitlab/config.py]: Done this.  (is now named "helper:")

File: gitlab/tests/test_config.py
Comment (klorenz) [reply gitlab/tests/test_config.py]: Improved the test and adapted to changed helper.

File: gitlab/config.py
Comment (klorenz) [reply gitlab/config.py]: Removed the ``shell=True`` and instead allow only a path to the helper program.  I also updated the documentation.

Comment (nejch) [reply gitlab/config.py]: Thank you for the changes @klorenz and sorry for the delay! I'll be merging in a minute.

I really liked the convenience of your initial implementation, and I don't want to block this PR any longer so I just pushed a small change (I credited you as the author still) - I went for a compromise where `shell=False` is still used and recommend wrappers for any scripts that require piping etc. I hope this is ok with you. 
I also expanded the docs and wrapped the call in a try/except block for people to have a meaningful erorr message on the CLI. See https://github.com/python-gitlab/python-gitlab/pull/1359/commits/91ffb8e97e213d2f14340b952630875995ecedb2, let me know if you want to revisit this in a follow-up.

**Review by nejch — APPROVED** [review]

**Comment by klorenz** [comment]
Hi @nejch, yes I am fine with your changes. This is a good compromise
between convenience and security. Thanks :)

Nejc Habjan ***@***.***> schrieb am So., 18. Apr. 2021, 12:11:

**Pull request merged**
